infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -226,49 +256,89 @@ public LogReplicationEntry receive(LogReplicationEntry message) {
             return null;
         }
 
+        return processReceivedMessage(message);
+    }
+
+    /**
+     * Process received (valid) message depending on the current rx state (LOG_ENTRY_SYNC or SNAPSHOT_SYNC)
+     *
+     * @param message received message
+     * @return ack
+     */
+    private LogReplicationEntry processReceivedMessage(LogReplicationEntry message) {
         if (rxState.equals(RxState.LOG_ENTRY_SYNC)) {
             return logEntrySinkBufferManager.processMsgAndBuffer(message);
         } else {
-            return snapshotSinkBufferManager.processMsgAndBuffer(message);
+            LogReplicationEntry ack = snapshotSinkBufferManager.processMsgAndBuffer(message);
+            if (ack.getMetadata().getMessageMetadataType() == MessageType.SNAPSHOT_REPLICATED) {
+                long lastStartedSnapshotSync = logReplicationMetadataManager.getLastSnapStartTimestamp();
+                // Verify this snapshot ACK corresponds to the last initialized/valid snapshot sync
+                // as a previous one could have been canceled but still processed due to messages being out of order
+                if (ack.getMetadata().getSnapshotTimestamp() == lastStartedSnapshotSync) {
+                    // Notify end of snapshot sync. This is a blocking call.
+                    snapshotSyncPlugin.onSnapshotSyncEnd(runtime);
+                } else {
+                    log.warn("SNAPSHOT_SYNC has completed for {}, but new ongoing SNAPSHOT_SYNC is {}",
+                            ack.getMetadata().getSnapshotTimestamp(), lastStartedSnapshotSync);
```

Comment:
ongoing snapshot sync is ack.getMetadata().getSnapshotTimestamp(), right?

Reply:
No, the ongoing one is the one persisted in the datastore. The ack might come from processing some old stuff in the buffer.